subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/PmdPluginIncrementalAnalysisIntegrationTest.groovy

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.plugins.quality
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.util.TestPrecondition
+
+class PmdPluginIncrementalAnalysisIntegrationTest extends AbstractIntegrationSpec {
+    def setup() {
+        buildFile << """
+            apply plugin: "java"
+            apply plugin: "pmd"
+
+            ${mavenCentralRepository()}
+
+            ${!TestPrecondition.FIX_TO_WORK_ON_JAVA9.fulfilled ? "sourceCompatibility = 1.6" : ""}
+        """.stripIndent()
+    }
+
+    def "incremental analysis cache file is not generated by default"() {
+        goodCode()
+
+        expect:
+        succeeds("check")
+        !file("build/pmd-cache").exists()
+    }
+
+    def "incremental analysis can be enabled"() {
+        goodCode()
+        buildFile << '''
+        pmd {
+            incrementalAnalysis = true
+        }
+'''
+
+        expect:
+        succeeds("check")
+        file("build/pmd-cache/pmdMain.cache").exists()
+    }
+
```

Comment:
I think we want a couple of more tests here (we don't need to go in-depth with how PMD works, but we should check a few use cases).  Is there a way to tell if PMD actually did incremental analysis (logging statements, particular output files, etc)? 

e.g.:
- `gradle pmdMain` then `gradle cleanPmdMain` then `gradle pmdMain`.  Do I have reports when incremental analysis is done?
- `gradle pmdMain` with one good file and one bad file.  Then delete the bad file, then `gradle pmdMain`.  Do I still have the bad file in the report?
- I think we'd want a test similar to `PmdPluginAuxclasspathIntegrationTest` where we demonstrate changes to a custom rule invalidate incremental analysis too.

Reply:
@big-guy feel free yo rely on us for expected behavior.

> gradle pmdMain then gradle cleanPmdMain then gradle pmdMain. Do I have reports when incremental analysis is done?

Yes you do. The results are the exact same with or without incremental analysis, we simply skip all parsing / rule applications since there are not changes to your sources / dependencies.

> gradle pmdMain with one good file and one bad file. Then delete the bad file, then gradle pmdMain. Do I still have the bad file in the report?

No you don't. Once again, with incremental analysis, the results should be the exact same as without it. Just a lot faster.

> I think we'd want a test similar to PmdPluginAuxclasspathIntegrationTest where we demonstrate changes to a custom rule invalidate incremental analysis too.

That's actually an interesting scenario. We do check the ruleset for changes, and the PMD version, but if additional Java rules change in the classpath we currently don't detect it. I'm tracking this as https://github.com/pmd/pmd/issues/603

Reply:
> Is there a way to tell if PMD actually did incremental analysis (logging statements, particular output files, etc)?

Yes there is. If enabled, PMD will use incremental analysis unless it finds any reason to invalidate the cache. When it does, it logs a message in the format `"Analysis cache invalidated, ${reason}"`.

For instance:
[`"Analysis cache invalidated, rulesets changed."`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java#L83)
[`"Analysis cache invalidated, classpath changed."`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cache/AbstractAnalysisCache.java#L96)
[`"Analysis cache invalidated, PMD version changed."`](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cache/FileAnalysisCache.java#L74)

Other reasons (ie: corrupted file, file permissions, etc.) currently don't follow this convention, but in retrospective it would be interesting to do so.

https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cache/FileAnalysisCache.java#L77-L79

Reply:
@jsotuyod However, I only saw the log when cache is invalidated, but I didn't see any log when cache is valid with `DEBUG` level.  Looks like you should log something like "using cache" or "cache file loaded" with at least `DEBUG` level here: https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cache/FileAnalysisCache.java#L71

Reply:
> gradle pmdMain then gradle cleanPmdMain then gradle pmdMain. Do I have reports when incremental analysis is done?
@big-guy Pmd lacks debug info (gradle pmdMain then gradle cleanPmdMain then gradle pmdMain. Do I have reports when incremental analysis is done?)